File: packages/web/src/payload.ts
export type PushIncomingPayload = {
  title?: string;
  body?: string;
  url?: string;
  icon?: string;
  image?: string;
  logId?: string;
  clickReportUrl?: string;
  deliveryReportUrl?: string;
  data?: Record<string, unknown>;
};

function asRecord(value: unknown): Record<string, unknown> | undefined {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }
  return undefined;
}

function pickString(record: Record<string, unknown>, keys: string[]): string | undefined {
  for (const key of keys) {
    const value = record[key];
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return undefined;
}

export function extractLogId(data: Record<string, unknown>): string | undefined {
  const direct = pickString(data, ["log_id", "logId", "push_id", "pushId"]);
  if (direct) return direct;
  for (const value of Object.values(data)) {
    if (typeof value === "string" && value.includes("log_id=")) {
      const match = value.match(/log_id=([^&]+)/);
      if (match?.[1]) return decodeURIComponent(match[1]);
    }
  }
  return undefined;
}

export function parsePushPayload(raw: unknown): PushIncomingPayload {
  const record = asRecord(raw);
  if (!record) return {};

  const data = asRecord(record.data) ?? {};
  const mergedData = { ...data };
  for (const [key, value] of Object.entries(record)) {
    if (!["data", "title", "body", "url", "icon", "image"].includes(key)) {
      mergedData[key] = value;
    }
  }

  const logId = extractLogId(mergedData) ?? pickString(record, ["log_id", "logId"]);
  const url = pickString(record, ["url"]) ?? pickString(mergedData, ["url"]);

  return {
    title: pickString(record, ["title"]),
    body: pickString(record, ["body"]),
    url,
    icon: pickString(record, ["icon"]),
    image: pickString(record, ["image"]),
    logId,
    clickReportUrl: pickString(mergedData, ["click_report_url", "clickReportUrl"]),
    deliveryReportUrl: pickString(mergedData, ["delivery_report_url", "deliveryReportUrl"]),
    data: mergedData,
  };
}

File: packages/flutter/lib/src/payload.dart
/// Parsed incoming push notification payload.
class PushIncomingPayload {
  const PushIncomingPayload({
    this.title,
    this.body,
    this.url,
    this.icon,
    this.image,
    this.logId,
    this.clickReportUrl,
    this.deliveryReportUrl,
    this.data = const {},
  });

  final String? title;
  final String? body;
  final String? url;
  final String? icon;
  final String? image;
  final String? logId;
  final String? clickReportUrl;
  final String? deliveryReportUrl;
  final Map<String, dynamic> data;
}

String? _pickString(Map<String, dynamic> map, List<String> keys) {
  for (final key in keys) {
    final value = map[key];
    if (value is String && value.trim().isNotEmpty) return value.trim();
  }
  return null;
}

String? extractLogId(Map<String, dynamic> data) {
  final direct = _pickString(data, ['log_id', 'logId', 'push_id', 'pushId']);
  if (direct != null) return direct;
  for (final value in data.values) {
    if (value is String && value.contains('log_id=')) {
      final match = RegExp(r'log_id=([^&]+)').firstMatch(value);
      if (match != null) return Uri.decodeComponent(match.group(1)!);
    }
  }
  return null;
}

PushIncomingPayload parsePushPayload(Map<String, dynamic> raw) {
  final data = Map<String, dynamic>.from(raw);
  final logId = extractLogId(data);
  return PushIncomingPayload(
    title: _pickString(data, ['title']),
    body: _pickString(data, ['body']),
    url: _pickString(data, ['url']),
    icon: _pickString(data, ['icon']),
    image: _pickString(data, ['image']),
    logId: logId,
    clickReportUrl: _pickString(data, ['click_report_url', 'clickReportUrl']),
    deliveryReportUrl:
        _pickString(data, ['delivery_report_url', 'deliveryReportUrl']),
    data: data,
  );
}
